internal/engine/compiler/engine.go

```diff
@@ -1009,13 +1010,11 @@ entry:
 				ce.builtinFunctionFunctionListenerBefore(ce.ctx, m, caller)
 			case builtinFunctionIndexFunctionListenerAfter:
 				ce.builtinFunctionFunctionListenerAfter(ce.ctx, m, caller)
-			case builtinFunctionIndexCheckExitCode:
-				// Note: this operation must be done in Go, not native code. The reason is that
-				// native code cannot be preempted and that means it can block forever if there are not
-				// enough OS threads (which we don't have control over).
-				if err := m.FailIfClosed(); err != nil {
-					panic(err)
-				}
+			case builtinFunctionIndexExitUnconditionally:
+				// Leave with the given exit code: the check has been performed in native code.
+				closed := atomic.LoadUint64(&m.Closed)
```

Comment:
this is likely just masking the underlying issue: it still fails randomly with `-race` (not surprising, since it's failing with `t.Parallel()`)

Reply:
and the root cause seems to be exactly read/writes to this field

(prev run: https://github.com/tetratelabs/wazero/actions/runs/4831243660/jobs/8608429404)

Reply:
actually, I may be wrong, because inspecting that further, it kind of looks like the failure is at `wasm.(*ModuleInstance).closeModuleOnCanceledOrTimeout()`; in that case that would mean that the race is on the `select/case` ? is that even possible, or is the error misleading?